impeller/docs/ubo_gles2.md

```diff
@@ -0,0 +1,24 @@
+# How Impeller Works Around The Lack of Uniform Buffers in Open GL ES 2.0.
+
+The Impeller Renderer API allows callers to specify uniform data in a discrete buffer. Indeed, it is conventional for the higher level layers of the Impeller stack to specify all uniform data for a render pass in a single buffer. This jumbo buffer is allocated using a simple bump allocator on the host before being transferred over to VRAM. The ImpellerC reflection engine generates structs with the correct padding and alignment such that the caller can just populate uniform struct members and `memcpy` them into the jumbo buffer, or use placement-new. Placement-new is used in cases where device buffers can be memory mapped into the client address space.
+
+This works extremely well when using a modern rendering backend like Metal. However, OpenGL ES 2.0 does not support uniform buffer objects. Instead, uniform data must be specified to GL from the client side using the `glUniform*` family of APIs. This poses a problem for the OpenGL backend implementation. From a view (an offset and range) into a buffer pointing to uniform data, it must infer the right uniform locations within a program object and bind uniform data at the right offsets within the buffer view.
+
+Since command generation is strongly typed, a pointer to metadata about the uniform information is stashed along with the buffer view in the command stream. This metadata is generated by the offline reflection engine part of ImpellerC. The metadata is usually a collection of items the runtime would need to infer the right `glUniform*` calls. An item in this collection would look like the following:
+
+```
+struct ShaderStructMemberMetadata {
+  ShaderType type; // the data type (bool, int, float, etc.)
+  std::string name; // the uniform member name "frame_info.mvp"
+  size_t offset;
+  size_t size;
+};
+```
+
+Using this mechanism, the runtime knows how to specify data from a buffer view to GL. But, this is still not sufficient as the buffer bindings are not known until after program link time.
```

Comment:
If I understand correctly, we use this metadata to map the jumbo buffer to the uniform names, and at encoding time we lookup the uniform locations using the names.
So my nit: Maybe "pull the uniforms out of the buffer view" would be slightly clearer than "specify data from a buffer view to GL". And "binding order" instead of "buffer bindings".

Reply:
Although I guess the metadata reference really is declaring the uniform layout, so maybe it's concise enough as-is.